File: 2b/ts/main.ts
const fs = require("fs");

const values = fs.readFileSync("input.txt").toString().split("\n");

class Position {
  x: number;
  y: number;
  aim: number;

  // Normal signature with defaults
  constructor(x = 0, y = 0, aim = 0) {
    this.x = x;
    this.y = y;
    this.aim = aim;
  }

  plotCourse(direction: string, magnitude: number) {
    switch (direction) {
      case "down":
        this.aim += magnitude;
        break;
      case "up":
        this.aim -= magnitude;
        break;
      case "forward":
        this.x += magnitude;
        this.y += this.aim * magnitude;
        if (this.y < 0) {
          this.y = 0;
        }
        break;
      default:
        console.log("Failed to parse direction");
        break;
    }
  }
}
let sub = new Position();

for (let i = 0; i < values.length; i++) {
  let tokens = values[i].split(" ");
  if (tokens[0] != "" && tokens[1] != "") {
    sub.plotCourse(tokens[0], parseInt(tokens[1]));
  }
}
console.log(sub);
console.log(sub.x * sub.y);

File: 2b/go/main.go
package main

import (
	"bufio"
	"log"
	"os"
	"strconv"
	"strings"
)

type position struct {
	x   int
	y   int
	aim int
}

func main() {
	readFile, err := os.Open("input.txt")

	if err != nil {
		log.Fatalf("failed to open file: %s", err)
	}

	fileScanner := bufio.NewScanner(readFile)
	fileScanner.Split(bufio.ScanLines)
	var directions []string
	var magnitudes []int

	for fileScanner.Scan() {
		tokens := strings.Split(fileScanner.Text(), " ")
		if len(tokens) == 2 {
			if tokens[0] != "" {
				directions = append(directions, tokens[0])
			} else {
				log.Printf("Failed to parse the direction. %v", tokens)
			}

			if tokens[1] != "" {
				num, err := strconv.Atoi(tokens[1])
				if err != nil {
					log.Printf("failed to convert to integer: %s", err)
				} else {
					magnitudes = append(magnitudes, num)
				}
			} else {
				log.Printf("Failed to parse the magnitude")
			}
		}
	}

	readFile.Close()

	subPosition := position{}
	for i := 0; i < len(directions); i++ {
		subPosition.plotCourse(directions[i], magnitudes[i])
	}

	log.Printf("x: %d, y: %d, aim: %d. Mult: %d", subPosition.x, subPosition.y, subPosition.aim, subPosition.x*subPosition.y)
}

func (pos *position) plotCourse(dir string, mag int) {
	switch dir {
	case "up":
		pos.aim -= mag
	case "down":
		pos.aim += mag
	case "forward":
		pos.x += mag
		pos.y += mag * pos.aim
		if pos.y < 0 {
			pos.y = 0
		}
	default:
		log.Printf("direction unknown")
	}
}
